File: android/src/main/kotlin/com/example/google_cast/extensions/QueueItemsExtensions.kt
package com.example.google_cast.extensions

import android.media.MediaDescription
import android.media.session.MediaSession
import android.os.Build
import com.google.android.gms.cast.MediaInfo
import com.google.android.gms.cast.MediaQueueItem

object QueueItemBuilder {

    fun fromMap(map: Map<String, Any?>): MediaQueueItem? {
//        'activeTrackIds': activeTrackIds,
//        'autoplay': autoPlay,
//        'customData': customData,
//        'itemId': itemId,
//        'media': mediaInformation.toMap(),
//        'playbackDuration': playbackDuration?.inSeconds,
//        'preloadTime': preLoadTime?.inSeconds,
//        'startTime': startTime?.inSeconds,

        val activeTrackIds = map["activeTrackIds"] as List<*>?
        val id = map["itemId"]
        var mediaInfo: MediaInfo? =
            GoogleCastMediaInfo.fromMap(map["media"] as Map<String, Any?>) ?: return null
        var builder = MediaQueueItem.Builder(mediaInfo!!)
        return builder.build()

    }

    fun listFromMap(items: List<Map<String, Any?>>): List<MediaQueueItem> {
        val list = mutableListOf<MediaQueueItem>()
        for (item in items) {
            val queueItem = QueueItemBuilder.fromMap(item)
            if (queueItem != null)
                list.add(queueItem)
        }
        return  list

    }

}

File: ios/flutter_chrome_cast/Sources/flutter_chrome_cast/Extensions/QueueItemsExtensions.swift
//
//  QueueItemsExtensions.swift
//  google_cast
//
//  Created by LUIZ FELIPE ALVES LIMA on 30/06/22.
//

import Foundation
import GoogleCast

extension GCKMediaQueueItem {
    static func fromMap(_ args : Dictionary<String, Any>) -> GCKMediaQueueItem? {
        
        let activeTrackIds = args["activeTrackIds"] as? [NSNumber]
        let autoplay = args["autoplay"] as? Bool ?? true
        guard let mediaDict = args["media"] as? Dictionary<String, Any>,
              let mediaInformation = GCKMediaInformation.fromMap(mediaDict) else {
            return nil
        }
        let startTime = (args["startTime"] as? NSNumber)?.doubleValue ?? 0
        let preloadTime = (args["preloadTime"] as? NSNumber)?.doubleValue ?? 0
        let customData = args["customData"] as? Dictionary<String,Any>
        let item = GCKMediaQueueItem(mediaInformation: mediaInformation, autoplay: autoplay, startTime: startTime, preloadTime: preloadTime, activeTrackIDs: activeTrackIds, customData: customData)
        
        
        return item
        
    }
    
    func toMap() -> Dictionary<String,Any> {
        var dict = Dictionary<String,Any>()
        dict["itemId"]  = self.itemID
        if !(self.preloadTime.isNaN || self.preloadTime.isInfinite) {
            dict["preloadTime"] =  Int(self.preloadTime)
        }
        if !(self.startTime.isNaN || self.startTime.isInfinite){
            dict["startTime"] = Int(self.startTime)
        }
        
        dict["mediaInformation"]  = self.mediaInformation.toMap()
        dict["autoPlay"] = self.autoplay
        dict["activeTracksIds"] = self.activeTrackIDs
        if !(self.playbackDuration.isNaN || self.playbackDuration.isInfinite){
            dict["playbackDuration"]  = Int(self.playbackDuration)
        }
        return dict
    }
}
